Library/src/actuators/Rudder.cpp

```diff
@@ -0,0 +1,182 @@
+    inv = inverted;
+
+    setpoint = Scalar(0);
+    theta = Scalar(0);
+    this->rudder = rudder;
+    this->rudder->BuildGraphicalObject();
+}
+
+Rudder::~Rudder()
+{
+    if(rudder != NULL)
+        delete rudder;
+}
+
+ActuatorType Rudder::getType()
+{
+    return ActuatorType::RUDDER;
+}
+
+void Rudder::setSetpoint(Scalar s)
+{
+    if(inv) s *= Scalar(-1);
+    setpoint = std::max(std::min(s, maxAngle), -maxAngle);
+}
+
+Scalar Rudder::getSetpoint()
+{
+    return inv ? -setpoint : setpoint;
+}
+
+Scalar Rudder::getAngle()
+{
+    return theta;
+}
+
+void Rudder::Update(Scalar dt)
+{
+    //Update rudder angle
+    theta = setpoint;
+
+    if(attach != NULL)
+    {
+        Quaternion rudderRot(theta, 0, 0);
+
+        //Get transforms
+        Transform solidTrans = attach->getCGTransform();
+        // o2a is the transform of the actuator
+        Transform rudderTrans = attach->getOTransform() * o2a * Transform(rudderRot);
+        Vector3 relPos = rudderTrans.getOrigin() - solidTrans.getOrigin();
+
+        Vector3 velocity = rudderTrans.getBasis().transpose()*attach->getLinearVelocityInLocalPoint(relPos);
+
+        Scalar angle = atan2(velocity.getY(), velocity.getX());
+
+        // If rudder facing velocity, reverse equations
+        if (fabs(angle) > .5*M_PI)
+        {
+            velocity *= -1.;
+            angle = atan2(velocity.getY(), velocity.getX());
+        }
+        
+        Ocean* ocn = SimulationApp::getApp()->getSimulationManager()->getOcean();
+        if(ocn->IsInsideFluid(rudderTrans.getOrigin()) && !velocity.isZero())
+        {
+            // Calculate quadratic approximations for lift and drag
+
+            // Equations from:
+            // Engelhardtsen, Oystein. 3D AUV Collision Avoidance.
+            // MS thesis. Institutt for teknisk kybernetikk, 2007.
+
+            // Fluid density
+            Scalar rho = ocn->getLiquid().density;
+
+            // Speed
+            Scalar u = velocity.length();
+
+            // d = angle
+            Scalar du2 = angle * u*u;
+
+            // Approximated rudder parameter
+            Scalar X = .5 * area * rho * liftCoeff;
+
+            // Equation 3.2.19
+            Scalar lift = du2 * X;
+
+            // Equation 3.2.21
+            Scalar drag = angle * du2 * .5 * X;
+
+            // Drag is opposite to velocity
+            dragV = -velocity;
+            dragV.normalize() *= drag;
+
+            // Lift is restricted to XY plane
+            // liftV = (-vy, vx, 0)
+            liftV = Vector3(0., 0., -1.).cross(velocity);
+            liftV.normalize() *= lift;
+
+            // Torque vectors
+            Vector3 dragT = relPos.cross(rudderTrans.getBasis() * dragV);
+            Vector3 liftT = relPos.cross(rudderTrans.getBasis() * liftV);
+
+            // Apply forces and torques
+            attach->ApplyTorque(dragT);
+            attach->ApplyTorque(liftT);
+            attach->ApplyCentralForce(rudderTrans.getBasis() * dragV);
```

Comment:
I am a bit uncertain about the application of the forces here, and the torques especially. Could you have a look at that @patrykcieslak ? I was unsure if I should apply the force as well as the corresponding torque.

Reply:
It looks ok but why you are not applying the lift force also, only the drag? I guess you forgot the line because you apply torques generated by both forces.

Reply:
@patrykcieslak I was not applying it because I was unsure if this general way was the right way to go. I'll add the lift force also then!

Reply:
Fixed